idl/partition_checksum.idl.hh

```diff
@@ -95,11 +96,15 @@ enum class node_ops_cmd : uint32_t {
      repair_updater,
 };
 
+class node_ops_id final {
```

Comment:
This is enough for idl. No need to include repair/id.hh.

Reply:
In the case of tagged_uuid this is correct, the idl can do without including the c++ definition.
The question is whether we want that or not.

Cc @nyh

We need a clear policy here. A tug of war between different reviewers is pointless.

Reply:
C++ should include IDL, not vice-versa.

Reply:
@avikivity but undoing https://github.com/scylladb/scylladb/commit/1fda686f966ebe1ccb39c5c0e1e310184d80a3fa will break the idl sefl-sufficiency and require including c++ header files from c++ source files just for resolving internal idl dependencies, without the c++ code needing them.

For example, deleting the include of query-request.hh from idl/uuid.idl.hh
that is required for defining `using query_id = utils::tagged_uuid<struct query_id_tag>` for its generated functions, like `serialize<query_id>`, causes a ripple effect where we need to include query-request.hh from files that directly, or indirectly, include `idl/uuid.dist.hh`, as follows.

This creates a mess that is much harder to maintain.

```diff
diff --git a/idl/uuid.idl.hh b/idl/uuid.idl.hh
index a4df34936d..b7ab972259 100644
--- a/idl/uuid.idl.hh
+++ b/idl/uuid.idl.hh
@@ -8,7 +8,6 @@
 
 #include "utils/UUID.hh"
 #include "schema_fwd.hh"
-#include "query-request.hh"
 #include "locator/host_id.hh"
 
 namespace utils {
diff --git a/db/hints/sync_point.cc b/db/hints/sync_point.cc
index 0f40da240c..69abf1f349 100644
--- a/db/hints/sync_point.cc
+++ b/db/hints/sync_point.cc
@@ -14,6 +14,7 @@
 
 #include "db/hints/sync_point.hh"
 #include "sync_point.hh"
+#include "query-request.hh"
 #include "idl/hinted_handoff.dist.hh"
 #include "idl/hinted_handoff.dist.impl.hh"
 #include "utils/base64.hh"
diff --git a/test/boost/idl_test.cc b/test/boost/idl_test.cc
index 874da2dee0..86a616f79d 100644
--- a/test/boost/idl_test.cc
+++ b/test/boost/idl_test.cc
@@ -152,6 +152,7 @@ struct const_template_arg_test_object {
     }
 };
 
+#include "query-request.hh"
 #include "idl/idl_test.dist.hh"
 #include "idl/idl_test.dist.impl.hh"
 
diff --git a/test/raft/randomized_nemesis_test.cc b/test/raft/randomized_nemesis_test.cc
index 85eb10d28c..d5063411ed 100644
--- a/test/raft/randomized_nemesis_test.cc
+++ b/test/raft/randomized_nemesis_test.cc
@@ -22,6 +22,7 @@
 #include "raft/logical_clock.hh"
 #include "serializer.hh"
 #include "serializer_impl.hh"
+#include "query-request.hh"
 #include "idl/uuid.dist.hh"
 #include "idl/uuid.dist.impl.hh"
 
```

Reply:
We can add support for it in the idl compiler.
Yet, this is a simple example.  I believe that there are more complex dependencies of the serialization code on c++ structures.  Maybe we just need to address all of them one at a time so that there will an idl definition for every structure that goes on the wire

Reply:
Look at the generated `frozen_schema.dist.hh` code for example:
```c++
namespace ser {

template <>
struct serializer<canonical_mutation> {
  template <typename Output>
  static void write(Output& buf, const canonical_mutation& v);

  template <typename Input>
  static canonical_mutation read(Input& buf);

  template <typename Input>
  static void skip(Input& buf);
};
```

`read` result requires `canonical_mutation` to be defined in full.

Reply:
So include it where `frozen_schema.dist.hh` is used. If in some places you want to use `frozen_schema.dist.hh`  because of other classes there and do not need `canonical_mutation` then probably `canonical_mutation` should be in a separate idl unit.

Can you include c++ files in headers generated by other IDL compilers?

Reply:
> So include it where `frozen_schema.dist.hh` is used. If in some places you want to use `frozen_schema.dist.hh` because of other classes there and do not need `canonical_mutation` then probably `canonical_mutation` should be in a separate idl unit.

This is the way it used to be in the past.
I think it proved to be messy and hard to maintain.
When making a change to idl I had to spend hours to figure out what I need to include and where
since one needs to include c++ header due to indirect dependencies. To figure these out you need to read the idl code instead of the compiler doing the work for you.

I really don't see the value in including the dependencies manually.

> 
> Can you include c++ files in headers generated by other IDL compilers?

didn't try that.  The support for `include` in the idl compiler was meant for including either c++ headers or other idl headers.

Reply:
> > So include it where `frozen_schema.dist.hh` is used. If in some places you want to use `frozen_schema.dist.hh` because of other classes there and do not need `canonical_mutation` then probably `canonical_mutation` should be in a separate idl unit.
> 
> This is the way it used to be in the past. I think it proved to be messy and hard to maintain. When making a change to idl I had to spend hours to figure out what I need to include and where since one needs to include c++ header due to indirect dependencies. To figure these out you need to read the idl code instead of the compiler doing the work for you.

Yes, if you add one more class serialization to existing idl file you suddenly need to add the include that describe that class in all places that include that idl as well. But may be the correct conclusion to take out from it is that that class needs to be in a separate idl file to not bring unneeded dependencies everywhere?

> 
> I really don't see the value in including the dependencies manually.
> 
> > Can you include c++ files in headers generated by other IDL compilers?
> 
> didn't try that. The support for `include` in the idl compiler was meant for including either c++ headers or other idl headers.

It was meant to include other idl headers.

Reply:
> > > So include it where `frozen_schema.dist.hh` is used. If in some places you want to use `frozen_schema.dist.hh` because of other classes there and do not need `canonical_mutation` then probably `canonical_mutation` should be in a separate idl unit.
> > 
> > 
> > This is the way it used to be in the past. I think it proved to be messy and hard to maintain. When making a change to idl I had to spend hours to figure out what I need to include and where since one needs to include c++ header due to indirect dependencies. To figure these out you need to read the idl code instead of the compiler doing the work for you.
> 
> Yes, if you add one more class serialization to existing idl file you suddenly need to add the include that describe that class in all places that include that idl as well. But may be the correct conclusion to take out from it is that that class needs to be in a separate idl file to not bring unneeded dependencies everywhere?

How will defining each idl class in a separate file will help consuming the generated headers in c++?
If the dependency is internal i.e. idl A requires idl B and idl B requires c++ header b.hh, the
the c++ source that includes A.dist.hh will need to include b.hh even if B is defined in a file of its own

> 
> > I really don't see the value in including the dependencies manually.
> > > Can you include c++ files in headers generated by other IDL compilers?
> > 
> > 
> > didn't try that. The support for `include` in the idl compiler was meant for including either c++ headers or other idl headers.
> 
> It was meant to include other idl headers.

I implemented it and I meant it to support include of both idl generated headers AND c++ headers.
It seems like this outreached the original intention of idl.
But the current code we generate from the idl definition file creates the dependency on the c++ definition of the classes.

And still nobody gave a satisfactory answer why "one shalt never include c++ headers from the generated idl headers".

We shouldn't stick to legacy design religiously due just because, if that comes in the way of streamlined progress.

Reply:
> > > > So include it where `frozen_schema.dist.hh` is used. If in some places you want to use `frozen_schema.dist.hh` because of other classes there and do not need `canonical_mutation` then probably `canonical_mutation` should be in a separate idl unit.
> > > 
> > > 
> > > This is the way it used to be in the past. I think it proved to be messy and hard to maintain. When making a change to idl I had to spend hours to figure out what I need to include and where since one needs to include c++ header due to indirect dependencies. To figure these out you need to read the idl code instead of the compiler doing the work for you.
> > 
> > 
> > Yes, if you add one more class serialization to existing idl file you suddenly need to add the include that describe that class in all places that include that idl as well. But may be the correct conclusion to take out from it is that that class needs to be in a separate idl file to not bring unneeded dependencies everywhere?
> 
> How will defining each idl class in a separate file will help consuming the generated headers in c++? 
If a file want to serialize a canonical mutation it already has to know what canonical mutation is and have corresponding includes.

> If the dependency is internal i.e. idl A requires idl B and idl B requires c++ header b.hh, the the c++ source that includes A.dist.hh will need to include b.hh even if B is defined in a file of its own

It is fine to include idl files in other idl files if there is a dependency. But dependency on external c++ file should be outside of the idl. I cannot imagine a case where a code wants to serialize something it has no idea about and hence it has to already include a c++ definition of what it wants to serialize. The only tricky part left is that idl generated headers should go after those includes that define serializable structures.

> 
> > > I really don't see the value in including the dependencies manually.
> > > > Can you include c++ files in headers generated by other IDL compilers?
> > > 
> > > 
> > > didn't try that. The support for `include` in the idl compiler was meant for including either c++ headers or other idl headers.
> > 
> > 
> > It was meant to include other idl headers.
> 
> I implemented it and I meant it to support include of both idl generated headers AND c++ headers. It seems like this outreached the original intention of idl. But the current code we generate from the idl definition file creates the dependency on the c++ definition of the classes.
> 
> And still nobody gave a satisfactory answer why "one shalt never include c++ headers from the generated idl headers".
>

Because it means that you create a dependency between source files that should not be there. If a file does not include canonical mutation it means it does not want to serialize it and in turn it means it does not need the code that knows how to do so. By including canonical mutation in an idl file you create artificial source files dependencies. Compiler tries to help you by saying "hey, here you want to serialize something you did not use, probably you want to move this into a separate idl', but instead you hid the warning. It would be great if our generated serialization code would be instantiated only when used. Then we wouldn't have such problems.
 
> We shouldn't stick to legacy design religiously due just because, if that comes in the way of streamlined progress.

What progress was achieved here? Removing some explicit includes? But this comes with a price of unneeded dependencies. May be this is a small price to pay. I do not know.

Reply:
> > > > > So include it where `frozen_schema.dist.hh` is used. If in some places you want to use `frozen_schema.dist.hh` because of other classes there and do not need `canonical_mutation` then probably `canonical_mutation` should be in a separate idl unit.
> > > > 
> > > > 
> > > > This is the way it used to be in the past. I think it proved to be messy and hard to maintain. When making a change to idl I had to spend hours to figure out what I need to include and where since one needs to include c++ header due to indirect dependencies. To figure these out you need to read the idl code instead of the compiler doing the work for you.
> > > 
> > > 
> > > Yes, if you add one more class serialization to existing idl file you suddenly need to add the include that describe that class in all places that include that idl as well. But may be the correct conclusion to take out from it is that that class needs to be in a separate idl file to not bring unneeded dependencies everywhere?
> > 
> > 
> > How will defining each idl class in a separate file will help consuming the generated headers in c++?
> > If a file want to serialize a canonical mutation it already has to know what canonical mutation is and have corresponding includes.
> 
> > If the dependency is internal i.e. idl A requires idl B and idl B requires c++ header b.hh, the the c++ source that includes A.dist.hh will need to include b.hh even if B is defined in a file of its own
> 
> It is fine to include idl files in other idl files if there is a dependency. But dependency on external c++ file should be outside of the idl.

I beg to differ.  There's no merit in doing that.
Generating self-sufficient idl headers works great and I don't see what we lose in doing so.
I think it's a net win.

> I cannot imagine a case where a code wants to serialize something it has no idea about and hence it has to already include a c++ definition of what it wants to serialize. The only tricky part left is that idl generated headers should go after those includes that define serializable structures.

In the example I gave there are several classes in a single idl file indeed, so the c++ user may be interested in one of them but it got the other inadvertently.  BUT, if there is nesting and the c++ user wants the top level structure serialization, why should it care about the nested dependencies?

> 
> > > > I really don't see the value in including the dependencies manually.
> > > > > Can you include c++ files in headers generated by other IDL compilers?
> > > > 
> > > > 
> > > > didn't try that. The support for `include` in the idl compiler was meant for including either c++ headers or other idl headers.
> > > 
> > > 
> > > It was meant to include other idl headers.
> > 
> > 
> > I implemented it and I meant it to support include of both idl generated headers AND c++ headers. It seems like this outreached the original intention of idl. But the current code we generate from the idl definition file creates the dependency on the c++ definition of the classes.
> > And still nobody gave a satisfactory answer why "one shalt never include c++ headers from the generated idl headers".
> 
> Because it means that you create a dependency between source files that should not be there.

Sadly, the dependency IS already there whether you like it or not, due to the code we generate.
I think that if the `read()` calls would return an allocated `std::unique_ptr<foo>` we might get away with only forward declaring foo, but current `serializer::read()` returns the in-core structure hence the dependency.

> If a file does not include canonical mutation it means it does not want to serialize it and in turn it means it does not need the code that knows how to do so. By including canonical mutation in an idl file you create artificial source files dependencies. Compiler tries to help you by saying "hey, here you want to serialize something you did not use, probably you want to move this into a separate idl', but instead you hid the warning. It would be great if our generated serialization code would be instantiated only when used. Then we wouldn't have such problems.
> 
> > We shouldn't stick to legacy design religiously due just because, if that comes in the way of streamlined progress.
> 
> What progress was achieved here? Removing some explicit includes? But this comes with a price of unneeded dependencies. May be this is a small price to pay. I do not know.

Reply:
> > > > > > So include it where `frozen_schema.dist.hh` is used. If in some places you want to use `frozen_schema.dist.hh` because of other classes there and do not need `canonical_mutation` then probably `canonical_mutation` should be in a separate idl unit.
> > > > > 
> > > > > 
> > > > > This is the way it used to be in the past. I think it proved to be messy and hard to maintain. When making a change to idl I had to spend hours to figure out what I need to include and where since one needs to include c++ header due to indirect dependencies. To figure these out you need to read the idl code instead of the compiler doing the work for you.
> > > > 
> > > > 
> > > > Yes, if you add one more class serialization to existing idl file you suddenly need to add the include that describe that class in all places that include that idl as well. But may be the correct conclusion to take out from it is that that class needs to be in a separate idl file to not bring unneeded dependencies everywhere?
> > > 
> > > 
> > > How will defining each idl class in a separate file will help consuming the generated headers in c++?
> > > If a file want to serialize a canonical mutation it already has to know what canonical mutation is and have corresponding includes.
> > 
> > 
> > > If the dependency is internal i.e. idl A requires idl B and idl B requires c++ header b.hh, the the c++ source that includes A.dist.hh will need to include b.hh even if B is defined in a file of its own
> > 
> > 
> > It is fine to include idl files in other idl files if there is a dependency. But dependency on external c++ file should be outside of the idl.
> 
> I beg to differ. There's no merit in doing that. Generating self-sufficient idl headers works great and I don't see what we lose in doing so. I think it's a net win.

If self sufficient means it includes everything and anything that you just trade one problem to another.

> 
> > I cannot imagine a case where a code wants to serialize something it has no idea about and hence it has to already include a c++ definition of what it wants to serialize. The only tricky part left is that idl generated headers should go after those includes that define serializable structures.
> 
> In the example I gave there are several classes in a single idl file indeed, so the c++ user may be interested in one of them but it got the other inadvertently. BUT, if there is nesting and the c++ user wants the top level structure serialization, why should it care about the nested dependencies?

If it is nesting in idl it is nesting in c++ and it means that if the code knows about top level it known about nesting structs as well. There is no magic in idl generated code. The example you gave is a primary example. I had similar problem with raft. There were source files that wanted to serialize only part of raft types (to store them on a storage) and there were files that wanted to serialize all of them (for messaging). Some types that were not needed by storage had a lot of c++ dependencies. By including those in idl I would create hidden false dependencies, but instead we have ` idl/raft_storage.idl.hh` and `idl/raft.idl.hh` and not false dependencies. Storage include only one and messaging both.

> 
> > > > > I really don't see the value in including the dependencies manually.
> > > > > > Can you include c++ files in headers generated by other IDL compilers?
> > > > > 
> > > > > 
> > > > > didn't try that. The support for `include` in the idl compiler was meant for including either c++ headers or other idl headers.
> > > > 
> > > > 
> > > > It was meant to include other idl headers.
> > > 
> > > 
> > > I implemented it and I meant it to support include of both idl generated headers AND c++ headers. It seems like this outreached the original intention of idl. But the current code we generate from the idl definition file creates the dependency on the c++ definition of the classes.
> > > And still nobody gave a satisfactory answer why "one shalt never include c++ headers from the generated idl headers".
> > 
> > 
> > Because it means that you create a dependency between source files that should not be there.
> 
> Sadly, the dependency IS already there whether you like it or not, due to the code we generate. I think that if the `read()` calls would return an allocated `std::unique_ptr<foo>` we might get away with only forward declaring foo, but current `serializer::read()` returns the in-core structure hence the dependency.

That not a false dependency. If a source file want to serialize foo it needs to know what the `foo` is and hence it will include "foo.hh". The false dependency is when a file wants to serialize `foo`, but needs to include "bar.hh" to pass a compilation. You can hide the include in the idl, but then the file will depend on "bar.hh" for no reason. 

> 
> > If a file does not include canonical mutation it means it does not want to serialize it and in turn it means it does not need the code that knows how to do so. By including canonical mutation in an idl file you create artificial source files dependencies. Compiler tries to help you by saying "hey, here you want to serialize something you did not use, probably you want to move this into a separate idl', but instead you hid the warning. It would be great if our generated serialization code would be instantiated only when used. Then we wouldn't have such problems.
> > > We shouldn't stick to legacy design religiously due just because, if that comes in the way of streamlined progress.
> > 
> > 
> > What progress was achieved here? Removing some explicit includes? But this comes with a price of unneeded dependencies. May be this is a small price to pay. I do not know.

Reply:
> > > 
> > > I beg to differ. There's no merit in doing that. Generating self-sufficient idl headers works great and I don't see what we lose in doing so. I think it's a net win.
> > 
> > 
> > If self sufficient means it includes everything and anything that you just trade one problem to another.
> 
> What is the other problem then?

I described it multiple times through all my replies: over broad includes leading to c++ source files depending on headers they should not depend on. It happens without idl as well of course, but you propose to hide those included in generated files which means it will be even harder to find.

> 
> #pragma once protects against including the c++ header multiple times (due to nesting, etc.) I don't think optimizing multiple includes is worth the effort of maintaining the the dependencies manually...
>
It is not about multiple times. It is about including headers the code does not need just to satisfy idl dependencies instead of fixing those dependencies.
 
> > > > I cannot imagine a case where a code wants to serialize something it has no idea about and hence it has to already include a c++ definition of what it wants to serialize. The only tricky part left is that idl generated headers should go after those includes that define serializable structures.
> > > 
> > > 
> > > In the example I gave there are several classes in a single idl file indeed, so the c++ user may be interested in one of them but it got the other inadvertently. BUT, if there is nesting and the c++ user wants the top level structure serialization, why should it care about the nested dependencies?
> > 
> > 
> > If it is nesting in idl it is nesting in c++ and it means that if the code knows about top level it known about nesting structs as well. There is no magic in idl generated code. The example you gave is a primary example. I had similar problem with raft. There were source files that wanted to serialize only part of raft types (to store them on a storage) and there were files that wanted to serialize all of them (for messaging). Some types that were not needed by storage had a lot of c++ dependencies. By including those in idl I would create hidden false dependencies, but instead we have ` idl/raft_storage.idl.hh` and `idl/raft.idl.hh` and not false dependencies. Storage include only one and messaging both.
> 
> I'm not saying the dependencies can't be satisfied by the c++ layer. My point it is that doing so across the board hard to maintain.
> 
> I think that [1fda686](https://github.com/scylladb/scylladb/commit/1fda686f966ebe1ccb39c5c0e1e310184d80a3fa) shows it very well by the amount of includes saved by making the generated idl headers self-sufficient.

It just moves includes around. The part that moves idl includes into the only idl files that need them is a good change. Putting c++ headers there not so much. Lets look at the first modification there in `idl/read_command.idl.hh`:

```

 */

+#include "query-request.hh"

+#include "idl/keys.idl.hh"
+#include "idl/range.idl.hh"
+#include "idl/uuid.idl.hh"

class cql_serialization_format final {
    uint8_t protocol_version();
};
```

You added `#include "query-request.hh"` there. The generated file is used by `message/messaging_service.cc`. It already include the same file explicitly (twice!) because if needs to handle the type in the C++ code. What was the point of hiding the include in the idl?

> > > Sadly, the dependency IS already there whether you like it or not, due to the code we generate. I think that if the `read()` calls would return an allocated `std::unique_ptr<foo>` we might get away with only forward declaring foo, but current `serializer::read()` returns the in-core structure hence the dependency.
> > 
> > 
> > That not a false dependency. If a source file want to serialize foo it needs to know what the `foo` is and hence it will include "foo.hh". The false dependency is when a file wants to serialize `foo`, but needs to include "bar.hh" to pass a compilation. You can hide the include in the idl, but then the file will depend on "bar.hh" for no reason.
> 
> We can narrow down the included headers to focus on the on-the-wire structures.
>

Yes, that what one should do when code does not compile for some reason, not just put an include into idl to silence it.

> > > > If a file does not include canonical mutation it means it does not want to serialize it and in turn it means it does not need the code that knows how to do so. By including canonical mutation in an idl file you create artificial source files dependencies. Compiler tries to help you by saying "hey, here you want to serialize something you did not use, probably you want to move this into a separate idl', but instead you hid the warning. It would be great if our generated serialization code would be instantiated only when used. Then we wouldn't have such problems.
> > > > > We shouldn't stick to legacy design religiously due just because, if that comes in the way of streamlined progress.
> > > > 
> > > > 
> > > > What progress was achieved here? Removing some explicit includes? But this comes with a price of unneeded dependencies. May be this is a small price to pay. I do not know.
> 
> I spent literally hours figuring out what I need to include and where instead of including the required c++ headers in the idl headers I changed.

I am sorry to hear that. But if after those hours you found out that there is a false dependency why not fix it instead?

Reply:
> > > > I beg to differ. There's no merit in doing that. Generating self-sufficient idl headers works great and I don't see what we lose in doing so. I think it's a net win.
> > > 
> > > 
> > > If self sufficient means it includes everything and anything that you just trade one problem to another.
> > 
> > 
> > What is the other problem then?
> 
> I described it multiple times through all my replies: over broad includes leading to c++ source files depending on headers they should not depend on. It happens without idl as well of course, but you propose to hide those included in generated files which means it will be even harder to find.
> 
> > #pragma once protects against including the c++ header multiple times (due to nesting, etc.) I don't think optimizing multiple includes is worth the effort of maintaining the the dependencies manually...
> 
> It is not about multiple times. It is about including headers the code does not need just to satisfy idl dependencies instead of fixing those dependencies.

I think these would have been needed to be included anyhow at the top layer.
I agree we should reduce the dependencies like you said.
This could be achieved by breaking the idl.hh files into smaller units, each including only what it needs.
And then reducing the c++ counterpart headers they depend on to be focused on on-the-wire classes without the extra baggage.

> 
> > > > > I cannot imagine a case where a code wants to serialize something it has no idea about and hence it has to already include a c++ definition of what it wants to serialize. The only tricky part left is that idl generated headers should go after those includes that define serializable structures.
> > > > 
> > > > 
> > > > In the example I gave there are several classes in a single idl file indeed, so the c++ user may be interested in one of them but it got the other inadvertently. BUT, if there is nesting and the c++ user wants the top level structure serialization, why should it care about the nested dependencies?
> > > 
> > > 
> > > If it is nesting in idl it is nesting in c++ and it means that if the code knows about top level it known about nesting structs as well. There is no magic in idl generated code. The example you gave is a primary example. I had similar problem with raft. There were source files that wanted to serialize only part of raft types (to store them on a storage) and there were files that wanted to serialize all of them (for messaging). Some types that were not needed by storage had a lot of c++ dependencies. By including those in idl I would create hidden false dependencies, but instead we have ` idl/raft_storage.idl.hh` and `idl/raft.idl.hh` and not false dependencies. Storage include only one and messaging both.
> > 
> > 
> > I'm not saying the dependencies can't be satisfied by the c++ layer. My point it is that doing so across the board hard to maintain.
> > I think that [1fda686](https://github.com/scylladb/scylladb/commit/1fda686f966ebe1ccb39c5c0e1e310184d80a3fa) shows it very well by the amount of includes saved by making the generated idl headers self-sufficient.
> 
> It just moves includes around.

I think it also eliminated duplication where a header that was included by multiple c++ files
is now included once by the dist.hh generated header.

> The part that moves idl includes into the only idl files that need them is a good change. Putting c++ headers there not so much. Lets look at the first modification there in `idl/read_command.idl.hh`:
> 
> ```
> 
>  */
> 
> +#include "query-request.hh"
> 
> +#include "idl/keys.idl.hh"
> +#include "idl/range.idl.hh"
> +#include "idl/uuid.idl.hh"
> 
> class cql_serialization_format final {
>     uint8_t protocol_version();
> };
> ```
> 
> You added `#include "query-request.hh"` there. The generated file is used by `message/messaging_service.cc`. It already include the same file explicitly (twice!) because if needs to handle the type in the C++ code. What was the point of hiding the include in the idl?

The change wasa first, mechanical, step, just to make the idl generated headers self-sufficient.
Further optimization of the dependencies would be great but it can be done in a follow up.

> 
> > > > Sadly, the dependency IS already there whether you like it or not, due to the code we generate. I think that if the `read()` calls would return an allocated `std::unique_ptr<foo>` we might get away with only forward declaring foo, but current `serializer::read()` returns the in-core structure hence the dependency.
> > > 
> > > 
> > > That not a false dependency. If a source file want to serialize foo it needs to know what the `foo` is and hence it will include "foo.hh". The false dependency is when a file wants to serialize `foo`, but needs to include "bar.hh" to pass a compilation. You can hide the include in the idl, but then the file will depend on "bar.hh" for no reason.
> > 
> > 
> > We can narrow down the included headers to focus on the on-the-wire structures.
> 
> Yes, that what one should do when code does not compile for some reason, not just put an include into idl to silence it.
> 

We took care to minimize the included headers for newly added idl classes for various tagged_uuid:s.
like here where we introduce repair/id.hh rather than include the much larger repair/repair.hh

> > > > > If a file does not include canonical mutation it means it does not want to serialize it and in turn it means it does 
not need the code that knows how to do so. By including canonical mutation in an idl file you create artificial source files dependencies. Compiler tries to help you by saying "hey, here you want to serialize something you did not use, probably you want to move this into a separate idl', but instead you hid the warning. It would be great if our generated serialization code would be instantiated only when used. Then we wouldn't have such problems.
> > > > > > We shouldn't stick to legacy design religiously due just because, if that comes in the way of streamlined progress.
> > > > > 
> > > > > 
> > > > > What progress was achieved here? Removing some explicit includes? But this comes with a price of unneeded dependencies. May be this is a small price to pay. I do not know.
> > 
> > 
> > I spent literally hours figuring out what I need to include and where instead of including the required c++ headers in the idl headers I changed.
> 
> I am sorry to hear that. But if after those hours you found out that there is a false dependency why not fix it instead?

Like I said, the main goal was to make the idl headers self-sufficient, not to optimize the dependencies.
We should definitely do that as a follow up.

Reply:
If an idl file includes another idl file, then the generated headers should reflect that and also include the dependency.

Reply:
> If an idl file includes another idl file, then the generated headers should reflect that and also include the dependency.

I'm not sure what you mean.

Let's look at a concrete example.
Currently, idl/frozen_schema.idl.hh has the following includes:
```
#include "canonical_mutation.hh"
#include "schema_mutations.hh"
#include "frozen_schema.hh"

#include "idl/uuid.idl.hh"
```

and idl/uuid.idl.hh has:
```
#include "utils/UUID.hh"
#include "schema_fwd.hh"
#include "query-request.hh"
#include "locator/host_id.hh"
```

The generated build/*/gen/frozen_schema.dist.hh has:
```
#include "serializer.hh"

#include "canonical_mutation.hh"
#include "schema_mutations.hh"
#include "frozen_schema.hh"
#include "idl/uuid.dist.hh"
```

And respectoively, uuid.dist.hh has:
```
#include "serializer.hh"

#include "utils/UUID.hh"
#include "schema_fwd.hh"
#include "query-request.hh"
#include "locator/host_id.hh"
```

So the generated idl/frozen_schema.dist.hh includes its own direct dependencies and the generated idl/uuid.dist.hh
and the latter includes its own direct dependencies.  The bottom line is the the higher level header nested dependencies are satisfies by the generated idl header files it includes.

Reply:
```
#include "canonical_mutation.hh"
#include "schema_mutations.hh"
#include "frozen_schema.hh"

#include "idl/uuid.idl.hh"
```

The first three includes violate the principle since they include stuff outside idl.

This is bad, since we might change something outside idl thinking it has no effect on compatibility, but it does.

Reply:
@avikivity, how would that affect compatibility?
The idl compiler doesn't actually include anything from those headers to generate the respective dist c++ files.
It just adds the include directives to those generated headers.
The same includes are and were needed anyhow before adding support for "include" in the idl files, so changing those headers will have exactly the same effect as it used to with the idl compiler adding these include directives.
It's just more manageable and defined in one place, instead of scattered all over the place for each .cc source file that includes a generated idl dist header.

Reply:
Well, it depends on whether we rely on definitions in that file for the wire format (for example using some_type = tagged_uuid<something>) there, or if we just need them for the constructors or accessors.

In any case, it's out of scope here. I'll review again.